python/tests/compat/test_dict.py

```diff
@@ -0,0 +1,217 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @file
+# @author Neil Vaytet
+
+import numpy as np
+import scipp as sc
+
+
+def test_variable_to_dict():
+    var = sc.Variable(dims=["x"],
+                      values=np.arange(10.),
+                      variances=np.random.random(10),
+                      unit=sc.units.m)
+    var_dict = sc.to_dict(var)
+    assert var_dict["dims"] == var.dims
+    assert var_dict["shape"] == var.shape
+    assert np.array_equal(var_dict["values"], var.values)
+    assert np.array_equal(var_dict["variances"], var.variances)
+    assert var_dict["unit"] == var.unit
+    assert var_dict["dtype"] == var.dtype
+
+
+def test_variable_from_dict():
```

Comment:
How do converters handle potentially missing mandatory keys from the dictionary. Do we get sensible errors?

Reply:
So if the keys are not right to create either a Variable or a DataArray, by default it tries to create a Dataset (since the keys of a Dataset could be anything).

When it then tries to create a DataArray for each key in the supplied dict (Dataset to be), I now added the error message if a DataArray cannot be created from it
```Py
raise KeyError("To create a DataArray, the supplied dict must contain "
                       "either 'data' or 'unaligned'. "
                       "Got {}.".format(d.keys()))
```

Reply:
> So if the keys are not right to create either a Variable or a DataArray, by default it tries to create a Dataset (since the keys of a Dataset could be anything).

I think this support what I hint at above: The dict created from a `Dataset` should not be a dict of data array dicts, rather: `{'data':dict(), 'coords':dict(), ...}`.